### PR TITLE
HMOS DPreference Library

### PR DESCRIPTION
**Pull request by arpithakori**

### CODE REVIEW

File: dpreference/src/test/java/ledongli/dpreference/ExampleUnitTest.java [hunk 1]
```diff
@@ -9,7 +9,7 @@
  */
 public class ExampleUnitTest {
     @Test
-    public void addition_isCorrect() throws Exception {
+    public void additionisCorrect() {
```
Comment (samuel-abhishek) [inline dpreference/src/test/java/ledongli/dpreference/ExampleUnitTest.java]: Test case not relevant to the library. Please delete the file.

File: entry/src/main/java/me/dozen/sample/Test.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
+package me.dozen.sample;
+
+import ohos.app.Context;
+import ohos.hiviewdfx.HiLog;
+import ohos.hiviewdfx.HiLogLabel;
+import me.dozen.dpreference.DPreference;
+
+/**
+ * Created by wangyida on 15/12/18.
+ */
```
Comment (samuel-abhishek) [inline entry/src/main/java/me/dozen/sample/Test.java:10]: This java doc can be removed as the file is developed from scratch

File: entry/src/test/java/ledongli/safepreference/ExampleUnitTest.java [hunk 3]
```diff
@@ -2,16 +2,14 @@
...
-
 import static org.junit.Assert.*;
 
 /**
  * To work on unit tests, switch the Test Artifact in the Build Variants view.
  */
 public class ExampleUnitTest {
     @Test
-    public void addition_isCorrect() throws Exception {
+    public void additionisCorrect() {
```
Comment (samuel-abhishek) [inline entry/src/test/java/ledongli/safepreference/ExampleUnitTest.java]: Test case not relevant to library. Please remove

File: dpreference/src/main/java/me/dozen/dpreference/PrefAccessor.java [hunk 4]
```diff
@@ -1,94 +1,166 @@
...
     }
 
     public static int getInt(Context context, String name, String key, int defaultValue) {
-        Uri URI = PreferenceProvider.buildUri(name, key, PreferenceProvider.PREF_INT);
+        Uri uri = PreferenceProvider.buildUri(name, key, PreferenceProvider.PREF_INT);
         int value = defaultValue;
-        Cursor cursor = context.getContentResolver().query(URI, null, null, null, null);
-        if (cursor != null && cursor.moveToFirst()) {
-            value = cursor.getInt(cursor.getColumnIndex(PreferenceProvider.PREF_VALUE));
+        DataAbilityPredicates dataAbilityPredicates = new DataAbilityPredicates((java.lang.String) null);
```
Comment (samuel-abhishek) [inline dpreference/src/main/java/me/dozen/dpreference/PrefAccessor.java:46]: null values need not be casted. Please check and remove the cast. Check this in all the methods

File: dpreference/src/main/java/me/dozen/dpreference/PrefAccessor.java [hunk 5]
```diff
@@ -1,94 +1,166 @@
...
+        Uri uri = PreferenceProvider.buildUri(name, key, PreferenceProvider.PREF_STRING);
+        ValuesBucket cv = new ValuesBucket();
+        cv.putString(PreferenceProvider.PREF_KEY, key);
+        cv.putString(PreferenceProvider.PREF_VALUE, value);
+        DataAbilityPredicates dataAbilityPredicates = new DataAbilityPredicates((java.lang.String) null);
+        dataAbilityPredicates.setWhereArgs(Arrays.asList());
+        try {
+            DataAbilityHelper.creator(context).update(uri, cv, dataAbilityPredicates);
+        } catch (DataAbilityRemoteException e) {
+            HiLog.error(HILOG_LABEL, "exception in getIntFromAttr");
```
Comment (samuel-abhishek) [inline dpreference/src/main/java/me/dozen/dpreference/PrefAccessor.java]: please use the method name in the log statement, to make it easier to identify during developer debug.

File: dpreference/src/main/java/me/dozen/dpreference/PreferenceProvider.java [hunk 6]
```diff
@@ -79,33 +90,30 @@ public Cursor query(Uri uri, String[] projection, String selection, String[] sel
...
     @Override
-    public int delete(Uri uri, String selection, String[] selectionArgs) {
-        switch (sUriMatcher.match(uri)) {
+    public int delete(Uri uri, DataAbilityPredicates predicates) {
+        String lastPath = PreferenceProvider.changeParamToPath(uri);
+        switch (getPathId(lastPath)) {
             case PREF_BOOLEAN:
             case PREF_LONG:
             case PREF_STRING:
             case PREF_INT:
```
Comment (samuel-abhishek) [inline dpreference/src/main/java/me/dozen/dpreference/PreferenceProvider.java]: Ideally, the preference must be deleted when delete api is called. But here removePreference implementation is missing in this delete api method. Please check.

File: dpreference/src/main/java/me/dozen/dpreference/PreferenceProvider.java [hunk 7]
```diff
@@ -114,12 +122,11 @@ public int delete(Uri uri, String selection, String[] selectionArgs) {
     }
 
     @Override
-    public int update(Uri uri, ContentValues values, String selection, String[] selectionArgs) {
+    public int update(Uri uri, ValuesBucket values, DataAbilityPredicates predicates) {
         PrefModel model = getPrefModelByUri(uri);
```
Comment (samuel-abhishek) [inline dpreference/src/main/java/me/dozen/dpreference/PreferenceProvider.java]: model variable could be null as getPrefModelByUri throws IllegalArgumentException in some scenarios. Please do a null check for model variable and surround it with try/catch. Do the same in all the places where getPrefModelByUri(uri) is called

File: dpreference/src/main/java/me/dozen/dpreference/PreferenceProvider.java [hunk 8]
```diff
@@ -1,67 +1,78 @@
...
     public static final String PREF_KEY = "key";
     public static final String PREF_VALUE = "value";
-
     public static final int PREF_BOOLEAN = 1;
     public static final int PREF_STRING = 2;
     public static final int PREF_INT = 3;
     public static final int PREF_LONG = 4;
-
-    private static final UriMatcher sUriMatcher;
+    public static final HiLogLabel HI_LOG_LABEL_2 = new HiLogLabel(0, 0, TAG);
```
Comment (samuel-abhishek) [inline dpreference/src/main/java/me/dozen/dpreference/PreferenceProvider.java]: Need not be public variable. Change it to private and remove the suffix "_2" from the variable name

**Review by samuel-abhishek — CHANGES_REQUESTED** [review]
Please handle the review comments.

File: dpreference/src/test/java/ledongli/dpreference/ExampleUnitTest.java (hunk 1 above)
Comment (arpithakori) [reply dpreference/src/test/java/ledongli/dpreference/ExampleUnitTest.java]: Done

File: entry/src/main/java/me/dozen/sample/Test.java (hunk 2 above)
Comment (arpithakori) [reply entry/src/main/java/me/dozen/sample/Test.java:10]: Done

File: entry/src/test/java/ledongli/safepreference/ExampleUnitTest.java (hunk 3 above)
Comment (arpithakori) [reply entry/src/test/java/ledongli/safepreference/ExampleUnitTest.java]: Done

File: dpreference/src/main/java/me/dozen/dpreference/PrefAccessor.java (hunk 4 above)
Comment (arpithakori) [reply dpreference/src/main/java/me/dozen/dpreference/PrefAccessor.java:46]: Done

File: dpreference/src/main/java/me/dozen/dpreference/PrefAccessor.java (hunk 5 above)
Comment (arpithakori) [reply dpreference/src/main/java/me/dozen/dpreference/PrefAccessor.java]: Done

File: dpreference/src/main/java/me/dozen/dpreference/PreferenceProvider.java (hunk 6 above)
Comment (arpithakori) [reply dpreference/src/main/java/me/dozen/dpreference/PreferenceProvider.java]: Done

File: dpreference/src/main/java/me/dozen/dpreference/PreferenceProvider.java (hunk 7 above)
Comment (arpithakori) [reply dpreference/src/main/java/me/dozen/dpreference/PreferenceProvider.java]: Done

File: dpreference/src/main/java/me/dozen/dpreference/PreferenceProvider.java (hunk 8 above)
Comment (arpithakori) [reply dpreference/src/main/java/me/dozen/dpreference/PreferenceProvider.java]: Done

**Review by samuel-abhishek — DISMISSED** [review]
LGTM

**Comment by samuel-abhishek** [comment]
Test this library on SDK6 and update the PR

**Review by samuel-abhishek — APPROVED** [review]
Okay

**Review by sharathks84 — APPROVED** [review]
okay approved

**Review by Gleelakrishna — APPROVED** [review]
OK

**Pull request merged**